src/core/thread/threadbase.d

```diff
@@ -0,0 +1,1374 @@
+class ThreadException : Exception
+{
+    @nogc @safe pure nothrow this(string msg, string file = __FILE__, size_t line = __LINE__, Throwable next = null)
+    {
+        super(msg, file, line, next);
+    }
+
+    @nogc @safe pure nothrow this(string msg, Throwable next, string file = __FILE__, size_t line = __LINE__)
+    {
+        super(msg, file, line, next);
+    }
+}
+
+
+/**
+* Base class for thread errors to be used for function inside GC when allocations are unavailable.
+*/
+class ThreadError : Error
+{
+    @nogc @safe pure nothrow this(string msg, string file = __FILE__, size_t line = __LINE__, Throwable next = null)
+    {
+        super(msg, file, line, next);
+    }
+
+    @nogc @safe pure nothrow this(string msg, Throwable next, string file = __FILE__, size_t line = __LINE__)
+    {
+        super(msg, file, line, next);
+    }
+}
+
+private
+{
+    // Handling unaligned mutexes are not supported on all platforms, so we must
+    // ensure that the address of all shared data are appropriately aligned.
+    import core.internal.traits : classInstanceAlignment;
+
+    enum mutexAlign = classInstanceAlignment!Mutex;
+    enum mutexClassInstanceSize = __traits(classInstanceSize, Mutex);
+
+    extern(C) void* swapContext(void* newContext) nothrow @nogc;
+
+    extern extern(C) immutable size_t threadSizeof;
+
+    extern(C) void* getStackBottom() nothrow @nogc;
+    extern(C) void* getStackTop() nothrow @nogc;
+}
+
+
+///////////////////////////////////////////////////////////////////////////////
+// Thread
+///////////////////////////////////////////////////////////////////////////////
+
+
+class ThreadBase
+{
+    ///////////////////////////////////////////////////////////////////////////
+    // Initialization
+    ///////////////////////////////////////////////////////////////////////////
+
+    this( void function() fn, size_t sz = 0 ) @safe pure nothrow @nogc
+    in( fn )
+    {
+        this(sz);
+        m_call = fn;
+    }
+
+    this( void delegate() dg, size_t sz = 0 ) @safe pure nothrow @nogc
+    in( dg )
+    {
+        this(sz);
+        m_call = dg;
+    }
+
+    /**
+     * Cleans up any remaining resources used by this object.
+     */
+    package bool destructBeforeDtor() nothrow @nogc
+    {
+        dataStorageDestroyIfAvail();
+
+        bool no_context = m_addr == m_addr.init;
+        bool not_registered = !next && !prev && (sm_tbeg !is this);
+
+        return (no_context || not_registered);
+    }
+
+    package void tlsGCdataInit() nothrow @nogc
+    {
+        m_tlsgcdata = rt_tlsgc_init();
+    }
+
+    package void dataStorageInit() nothrow
+    {
+        assert( m_curr is &m_main );
+
+        m_main.bstack = getStackBottom();
+        m_main.tstack = m_main.bstack;
+        tlsGCdataInit();
+    }
+
+    package void dataStorageDestroy() nothrow @nogc
+    {
+        rt_tlsgc_destroy( m_tlsgcdata );
+        m_tlsgcdata = null;
+    }
+
+    package void dataStorageDestroyIfAvail() nothrow @nogc
```

Comment:
The usual naming convention is to start with the verb, e.g. `destroyDataStorage`.